ENH: add support for reading .tar archives

Skn0tt opened the pull request:
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

At the moment, reading from `.tar.gz` will decode using `.gzip`, but interprets the `tar` contents as if it were a `csv`. This leads to funky behaviour (due to the way `tar` files are structured, it will interpret filenames as column names), and is generally incorrect.

At the moment, Pandas users can work-around this using `tarfile`:

```python
import pandas as pd
import tarfile

with tarfile.open("file.tar.gz", "r:*") as tar:
    file = tar.getnames()[0]
    return pd.read_csv(tar.extractfile(file))
```

This PR adds this logic into Pandas, similar to how Pandas already supports reading from `.zip` files.

Co-Authored-By: @margarete01

Skn0tt committed:
Add reproduction test for .tar.gz archives

co-authored-by: Margarete Dippel <margarete01@users.noreply.github.com>

Skn0tt committed:
add support for .tar archives

python's `tarfile` supports gzip, xz and bz2 encoding,
so we don't need to make any special cases for that.

co-authored-by: Margarete Dippel <margarete01@users.noreply.github.com>

Skn0tt committed:
update doc comments

pep8speaks commented:
Hello @Skn0tt! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/core/frame.py):

> [Line 3023:89](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/core/frame.py#L3023): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)
> [Line 3026:89](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/core/frame.py#L3026): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 88 characters)



* In the file [`pandas/io/json/_json.py`](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/io/json/_json.py):

> [Line 481:89](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/io/json/_json.py#L481): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)





* In the file [`pandas/io/xml.py`](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/io/xml.py):

> [Line 807:89](https://github.com/pandas-dev/pandas/blob/e6730613e279b86d4a673668421e6a947548c280/pandas/io/xml.py#L807): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 88 characters)

Skn0tt committed:
fix: pep8 errors

twoertwein commented on pandas/io/common.py:
```diff
@@ -520,6 +521,9 @@ def infer_compression(
             # Cannot infer compression of a buffer, assume no compression
             return None
 
+        if ".tar" in filepath_or_buffer:
```
There are only four options for tar (.tar, tar.gz, tar.bz, tar.xz), or? Might be simpler to add them to `_compression_to_extension`.

Skn0tt replied:
Makes sense! Implemented in https://github.com/pandas-dev/pandas/pull/44787/commits/6a8edef38437769491c645e284ec2696e13b8182 and https://github.com/pandas-dev/pandas/pull/44787/commits/d4e40c97b67ab0ad98e5435b0ef0710310fb032e.

twoertwein commented on pandas/io/common.py:
```diff
@@ -747,6 +751,21 @@ def get_handle(
 
+        # TAR Encoding
+        elif compression == "tar":
+            tar = tarfile.open(handle, "r:*")
```
If pandas supports reading from .tar*, users will probably also expect being able to write to .tar*.

Skn0tt replied:
Implemented in https://github.com/pandas-dev/pandas/pull/44787/commits/5f22df77accc57cfcd9864052db5b58e5dcae2e3 ✓

twoertwein commented on pandas/tests/io/parser/test_compression.py:
```diff
@@ -162,6 +162,14 @@ def test_invalid_compression(all_parsers, invalid_compression):
 
+@skip_pyarrow
+def test_compression_tar_archive(all_parsers, csv_dir_path):
+    parser = all_parsers
```
https://github.com/pandas-dev/pandas/blob/04b538a55365b5d1ae930687a91bb3ac4ea53040/pandas/conftest.py#L271
and
https://github.com/pandas-dev/pandas/blob/04b538a55365b5d1ae930687a91bb3ac4ea53040/pandas/conftest.py#L279

Skn0tt replied:
We decided to only add one user-facing compression option called `tar`, and have the `.gz` or `.bz2` or `.xz` thing inferred from the filename. If a developer wants to enforce a specific encoding independent from the filename, they can specify the compression by setting `mode="w:gz"`.

Skn0tt replied:
Added to the fixtures in https://github.com/pandas-dev/pandas/pull/44787/commits/5f22df77accc57cfcd9864052db5b58e5dcae2e3#diff-d591578c09ad62c8fb10500f126a3657c502db725ec6ec6d7cb9a6eb28ab9dca.

twoertwein replied:
Thanks, seems reasonable to me.

Skn0tt committed:
refactor: flip _compression_to_extension around to support multiple extensions on same compression

co-authored-by: Margarete Dippel <margarete01@users.norepl
y.github.com>

Skn0tt committed:
refactor: detect tar files using existing extension mapping

co-authored-by: Margarete Dippel <margarete01@users.noreply.github.com>

Skn0tt committed:
feat: add support for writing tar files

co-authored-by: Margarete Dippel <margarete01@users.noreply.github.com>

Skn0tt committed:
feat: assure it respects .gz endings

Skn0tt committed:
Merge branch 'master' into read-tar-archives

Skn0tt commented:
Hey @twoertwein, thank you for the review! Addressed the feedback. I think we're still missing some documentation things here and there.

- [ ] TODO: add `tar` as compression option to docs

Skn0tt committed:
feat: add "tar" entry to compressionoptions

Skn0tt committed:
chore: add whatsnew entry

Skn0tt committed:
fix: test_compression_size_fh

twoertwein commented on pandas/io/common.py:
```diff
@@ -823,6 +852,96 @@ def get_handle(
     )
 
 
+class _BytesTarFile(tarfile.TarFile, BytesIO):
```
Is it possible to share code with the _BytesZipFile class? Maybe _BytesCompressMixin from which both classes inherit?

twoertwein replied:
Can you comment on why we need this wrapper?

Skn0tt replied:
I tried extracting some of their code into a Mixin, but found that there's little room for abstraction. Except for the three lines of `write`, none of the methods are identical. Since adding a Mixin also makes the code harder to follow, I'd prefer to keep the duplication. Having said that, if you see a good way of abstracting here, I'm more than open to it!

Skn0tt replied:
Added a comment in 887fd10f80e380d7bd907894927a8a368556dce1.

twoertwein commented on pandas/core/generic.py:
```diff
@@ -2341,6 +2341,7 @@ def to_json(
         default_handler: Callable[[Any], JSONSerializable] | None = None,
         lines: bool_t = False,
         compression: CompressionOptions = "infer",
+        mode: str = "w",
```
Is mode needed? I think we expect/require that file handles are opened in binary mode when the user request compression.

Skn0tt replied:
Nope, not needed. I realised that the passthrough described in

https://github.com/pandas-dev/pandas/pull/44787/files#diff-132ee3be1f83a9f885442f45ed9ccbc96796ae28f97991b7c99ce25d44fd6af7R206

doesn't yet work. Fixed in https://github.com/pandas-dev/pandas/pull/44787/commits/57eba0ad7b63c984fb69c8c2a7e1a40ae3721bd6, and removed the added `mode` parameter.

twoertwein commented on pandas/core/frame.py:
```diff
@@ -3020,11 +3020,13 @@ def to_xml(
+        compression : {{'infer', 'gzip', 'bz2',
+            'zip', 'tar', 'xz', None}}, default 'infer'
             For on-the-fly decompression of on-disk data. If 'infer', then use
-            gzip, bz2, zip or xz if path_or_buffer is a string ending in
```
Don't worry about the doc-strings for now. #43925 creates one shared doc-string for all the compression doc-strings.

twoertwein replied:
#43925 was merged

jreback commented:
will need https://github.com/pandas-dev/pandas/pull/43925 and review

Skn0tt committed:
Merge branch 'master' into read-tar-archives

Skn0tt committed:
add tarfile to shared compression docs

Skn0tt committed:
fix formatting

Skn0tt commented:
Merged in #43925 and added to the shared docs.

Skn0tt committed:
pass through "mode" via compression args

Skn0tt committed:
fix pickle test

Skn0tt committed:
add class comment

jreback commented:
@Skn0tt if you can update here

Skn0tt commented:
couldn't find it :/ I'll dedicate some time this weekend to do that work again

Skn0tt committed:
Merge remote-tracking branch 'origin/main' into read-tar-archives

Skn0tt committed:
sort imports

Skn0tt committed:
add _compression_to_extension back for backwards compatibility

Skn0tt committed:
fix some type warnings

Skn0tt committed:
fix: formatting

Skn0tt committed:
fix: mypy complaints

Skn0tt committed:
fix: more tests

Skn0tt committed:
fix: some error with xml

Skn0tt committed:
fix: interpreted text role

Skn0tt committed:
move to v1.5 whatsnw

Skn0tt commented:
Tests are green, and all checks apart from one are green too. That last failing check is from a Azure Pipelines timeout - I‘m pretty confident that this goes away when re-run.

jreback commented on pandas/core/shared_docs.py:
```diff
@@ -421,12 +421,16 @@
+    ``zipfile.ZipFile``, ``gzip.GzipFile``,
+    ``bz2.BZ2File``, ``zstandard.ZstdDecompressor`` or
+    ``tarfile.TarFile``, respectively.
+    As an example, the following could be passed for faster compression and to create
```
add a versionadded 1.5 for the .tar change

Skn0tt replied:
addressed in addressed in https://github.com/pandas-dev/pandas/pull/44787/commits/9a4fa074f32916bbba4f2eb76be1ed6e2835feeb

jreback commented on pandas/core/shared_docs.py:
```diff
@@ -435,13 +439,17 @@
+    to one of {``'zip'``, ``'gzip'``, ``'bz2'``, ``'zstd'``, ``'tar'``} and other
+    key-value pairs are forwarded to
+    ``zipfile.ZipFile``, ``gzip.GzipFile``,
+    ``bz2.BZ2File``, ``zstandard.ZstdDecompressor`` or
```
same as above

Skn0tt replied:
addressed in https://github.com/pandas-dev/pandas/pull/44787/commits/9a4fa074f32916bbba4f2eb76be1ed6e2835feeb

jreback commented on pandas/io/common.py:
```diff
@@ -753,6 +761,30 @@ def get_handle(
+                    assert file is not None
+                    handle = file
+                elif len(files) == 0:
+                    raise ValueError(f"Zero files found in TAR archive {path_or_buf}")
```
test  hits this?

Skn0tt replied:
no, not yet. I'll add a test for this both for TAR and ZIP (which is currently also untested for this)

Skn0tt replied:
added test in https://github.com/pandas-dev/pandas/pull/44787/commits/086c59883e169fe021873f25304e7bb09f2823fd

jreback commented on pandas/io/common.py:
```diff
@@ -753,6 +761,30 @@ def get_handle(
+                else:
+                    raise ValueError(
+                        "Multiple files found in TAR archive. "
+                        f"Only one file per TAR archive: {files}"
```
test hits this?

Skn0tt replied:
same as above

Skn0tt replied:
added test in https://github.com/pandas-dev/pandas/pull/44787/commits/086c59883e169fe021873f25304e7bb09f2823fd

jreback commented on pandas/io/common.py:
```diff
@@ -844,6 +876,116 @@ def get_handle(
+# definition in base class "BytesIO"  [misc]
+# error: Definition of "read" in base class "TarFile" is incompatible with
+# definition in base class "IO"  [misc]
+class _BytesTarFile(tarfile.TarFile, BytesIO):  # type: ignore[misc]
```
umm can we inherit something similar to BytesZip and make this easier (or create a base class)

Skn0tt replied:
@twoertwein also [commented this](https://github.com/pandas-dev/pandas/pull/44787#discussion_r770164095). I found it hard to implement, as there's not as much shared logic between those classes as it might seem in first glance.

jreback reviewed (changes requested):
a few more comments

jreback commented:
cc @twoertwein if you can have a look.

Skn0tt committed:
add versionadded note

Skn0tt committed:
don't leave blank lines

twoertwein commented on pandas/_typing.py:
```diff
@@ -256,7 +256,7 @@ def closed(self) -> bool:
 CompressionDict = Dict[str, Any]
 CompressionOptions = Optional[
-    Union[Literal["infer", "gzip", "bz2", "zip", "xz", "zstd"], CompressionDict]
+    Union[Literal["infer", "gzip", "bz2", "zip", "xz", "zstd", "tar"], CompressionDict]
```
Does tar.gz, tar.z2, ... work?

Skn0tt replied:
yes! the `tarfile` standard library we're using has support for either `.gz`, `.z2` etc: https://docs.python.org/3/library/tarfile.html
since this `CompressionOptions` dict only specifies the pandas-internal decompression mechanism to use, we don't need to list any other extensions here.

Skn0tt committed:
add tests for zero files / multiple files

twoertwein commented on pandas/_testing/_io.py:
```diff
@@ -398,6 +400,14 @@ def write_to_compressed(compression, path, data, dest="test"):
         mode = "w"
         args = (dest, data)
         method = "writestr"
+    elif compression == "tar":
```
I think this entire function could be replaced with a call to `get_handle` (not needed in this PR)

twoertwein commented on pandas/io/common.py:
```diff
@@ -450,12 +452,20 @@ def file_path_to_url(path: str) -> str:
-    "zip": ".zip",
-    "xz": ".xz",
-    "zstd": ".zst",
+    value: key for key, value in _extension_to_compression.items()
```
This inverts `_extension_to_compression`? Will this not end up with `".tar": "tar.xz"` (overwriting the previous tar keys?)

Skn0tt replied:
that's true. `_extension_to_compression` is only used in tests as fixtures, though: https://cs.github.com/?scopeName=All+repos&scope=&q=repo%3Apandas-dev%2Fpandas+_compression_to_extension
since we don't need to test every single variant of `.tar` (that's dealed with by python's builtin `tarfile`), I reckon it's fine to only test for one of the variants.

twoertwein replied:
Would be nice to move it then to the tests (if it makes sense). Testing only one of them seems fine with me. Might need a test for the new mode arguments.

Skn0tt replied:
will do. with the `mode` argument, I'll see if there's a way to *not* add it - there's plenty of ways how users can misunderstand the argument, and adding it for the sake of different tar compression options doesn't seem like good enough of a reason

twoertwein replied:
Okay! When seeing `mode`, I mainly think about text vs binary. Some IO functions currently don't have it (technically all of the csv/json should have it - in case the text/binary auto-detection fails).

Skn0tt replied:
moved `_compression_to_extension` into tests in https://github.com/pandas-dev/pandas/pull/44787/commits/861faf0e3cf86374946b56767f43067da0fde349.

twoertwein commented on pandas/io/json/_json.py:
```diff
@@ -81,6 +81,7 @@ def to_json(
     default_handler: Callable[[Any], JSONSerializable] | None = None,
     lines: bool = False,
     compression: CompressionOptions = "infer",
+    mode: str = "w",
```
These new `mode` arguments might also need a whatsnew

twoertwein commented on pandas/io/pickle.py:
```diff
@@ -96,7 +97,7 @@ def to_pickle(
     with get_handle(
         filepath_or_buffer,
-        "wb",
+        mode,
```
Would it make sense to raise an error if "b" is not in mode?

Skn0tt replied:
I'd fear that's unreliable. the only real reason for people to pass in a different mode here, is if they want to use one of `tarfile`'s [custom modes](https://docs.python.org/3/library/tarfile.html#tarfile.open) like `w:gz`.

Skn0tt replied:
as an alternative, we could replace the `mode` argument with an argument that's specific to tarfile?

twoertwein replied:
If the user can specify mode as part of the `compression_options`, I would remove mode.

Skn0tt replied:
removed `mode` in https://github.com/pandas-dev/pandas/pull/44787/commits/9458ecbe116c83e04ff2bd111467caa45a16f650 and added a test ensuring that `compression.mode` can be used to specify compression method in https://github.com/pandas-dev/pandas/pull/44787/commits/d20f31505960515bc43fe0074e52ecd141f1940d

Skn0tt committed:
move _compression_to_extension to tests

Skn0tt committed:
revert added "mode" argument

Skn0tt committed:
add test to ensure that `compression.mode` works

twoertwein reviewed (approved):
Thank you @Skn0tt looks good to me!

I hope that the compression wrappers can be simplified in the future. Another future todo is to close file handles when the tar archive contains no/too many files (zip has the same issue).

mroeschke reviewed (commented):
Thanks for sticking with this. Could you merge main one more time? It looks fairly close to merging

Skn0tt committed:
Merge branch 'main' into read-tar-archives

mroeschke reviewed (changes requested):
Looks like some tests are still failing on Windows

Skn0tt committed:
Merge branch 'main' into read-tar-archives

Skn0tt committed:
compare strings, not bytes

Skn0tt committed:
replace carriage returns

mroeschke commented on pandas/tests/io/test_compression.py:
```diff
@@ -255,3 +269,61 @@ def test_bzip_compression_level(obj, method):
+                members = archive.getmembers()
+                assert len(members) == 1
+                content = archive.extractfile(members[0]).read().decode("utf8")
+                content = content.replace("\r\n", "\n")  # windows
```
Could you test this based on the platform instead? There is `pandas.compat.is_platform_windows`

jreback reviewed (approved):
looks reasonable. over to you @twoertwein & @mroeschke

mroeschke commented:
Thanks @Skn0tt. If you could follow up with this https://github.com/pandas-dev/pandas/pull/44787#discussion_r867014069= in another PR that would be great
